src/type.js

```diff
@@ -0,0 +1,312 @@
+import {
+  getConfig as getDOMTestingLibraryConfig,
+  fireEvent,
+} from '@testing-library/dom'
+import {tick} from './tick'
+
+function wait(time) {
+  return new Promise(resolve => setTimeout(() => resolve(), time))
+}
+
+// this needs to be wrapped in the asyncWrapper for React's act and angular's change detection
+async function type(...args) {
+  let result
+  await getDOMTestingLibraryConfig().asyncWrapper(async () => {
+    result = await typeImpl(...args)
+  })
+  return result
+}
+
+async function typeImpl(element, text, {allAtOnce = false, delay} = {}) {
+  if (element.disabled) return
+
+  element.focus()
+
+  // The focused element could change between each event, so get the currently active element each time
+  const currentElement = () => element.ownerDocument.activeElement
+  const currentValue = () => element.ownerDocument.activeElement.value
+
+  if (allAtOnce) {
+    if (!element.readOnly) {
+      fireEvent.input(element, {
+        target: {value: calculateNewValue(text)},
+      })
+    }
+  } else {
+    const eventCallbackMap = {
+      ...modifier({
+        name: 'shift',
+        key: 'Shift',
+        keyCode: 16,
+        modifierProperty: 'shiftKey',
+      }),
+      ...modifier({
+        name: 'ctrl',
+        key: 'Control',
+        keyCode: 17,
+        modifierProperty: 'ctrlKey',
+      }),
+      ...modifier({
+        name: 'alt',
+        key: 'Alt',
+        keyCode: 18,
+        modifierProperty: 'altKey',
+      }),
+      ...modifier({
+        name: 'meta',
+        key: 'Meta',
+        keyCode: 93,
+        modifierProperty: 'metaKey',
+      }),
+      '{enter}': async ({eventOverrides}) => {
+        const key = 'Enter'
+        const keyCode = 13
+
+        const keyDownDefaultNotPrevented = fireEvent.keyDown(currentElement(), {
+          key,
+          keyCode,
+          which: keyCode,
+          ...eventOverrides,
+        })
+
+        if (keyDownDefaultNotPrevented) {
+          await tick()
+
+          fireEvent.keyPress(currentElement(), {
+            key,
+            keyCode,
+            charCode: keyCode,
+            ...eventOverrides,
+          })
+        }
+
+        if (currentElement().tagName === 'BUTTON') {
```

Comment:
Good thinking here! I didn't even think about non-text elements. Now that I am though, would it be appropriate to also handle `<input type="button" />`, `<input type="submit" />`, etc?